File: day_04/day_04.py
data = []
str = "XMAS"

with open('input.txt', 'r') as file:
    data = [line.strip() for line in file]

# for i in range(len(data)):
#     print(data[i])

count = 0

def in_range(y, x):
    return (y >= 0 and y < len(data) and 
            x >= 0 and x < len(data[y]))

def check_xmas(y, x):
    global count

    if in_range(y - 1, x - 1) and data[y - 1][x - 1] == 'M':
       if in_range(y - 2, x - 2) and data[y - 2][x - 2] == 'A':
           if in_range(y - 3, x - 3) and data[y - 3][x - 3] == 'S':
               count += 1

    if in_range(y - 1, x) and data[y - 1][x] == 'M':
        if in_range(y - 2, x) and data[y - 2][x] == 'A':
            if in_range(y - 3, x) and data[y - 3][x] == 'S':
                count += 1
               
    if in_range(y - 1, x + 1) and data[y - 1][x + 1] == 'M':
        if in_range(y - 2, x + 2) and data[y - 2][x + 2] == 'A':
            if in_range(y - 3, x + 3) and data[y - 3][x + 3] == 'S':
                count += 1
               
    if in_range(y, x - 1) and data[y][x - 1] == 'M':
        if in_range(y, x - 2) and data[y][x - 2] == 'A':
            if in_range(y, x - 3) and data[y][x - 3] == 'S':
               count += 1
               
    if in_range(y, x + 1) and data[y][x + 1] == 'M':
        if in_range(y, x + 2) and data[y][x + 2] == 'A':
            if in_range(y, x + 3) and data[y][x + 3] == 'S':
                count += 1
               
    if in_range(y + 1, x - 1) and data[y + 1][x - 1] == 'M':
        if in_range(y + 2, x - 2) and data[y + 2][x - 2] == 'A':
            if in_range(y + 3, x - 3) and data[y + 3][x - 3] == 'S':
                count += 1
               
    if in_range(y + 1, x) and data[y + 1][x] == 'M':
       if in_range(y + 2, x) and data[y + 2][x] == 'A':
           if in_range(y + 3, x) and data[y + 3][x] == 'S':
               count += 1
               
    if in_range(y + 1, x + 1) and data[y + 1][x + 1] == 'M':
        if in_range(y + 2, x + 2) and data[y + 2][x + 2] == 'A':
            if in_range(y + 3, x + 3) and data[y + 3][x + 3] == 'S':
                count += 1


for y in range(len(data)):
    for x in range(len(data[y])):
        if data[y][x] == 'X':
            # print(y , x)
            # check_char(y, x, 1)
            check_xmas(y, x)

print("#1 ->", count)
count = 0

def check_x_mas(y, x):
    global count

    # top two
    if in_range(y - 1, x - 1) and data[y - 1][x - 1] == 'M' and in_range(y - 1, x + 1) and data[y - 1][x + 1] == 'M':
        if in_range(y + 1, x - 1) and data[y + 1][x - 1] == 'S' and in_range(y + 1, x + 1) and data[y + 1][x + 1] == 'S':
            count += 1

    # left two
    if in_range(y - 1, x - 1) and data[y - 1][x - 1] == 'M' and in_range(y + 1, x - 1) and data[y + 1][x - 1] == 'M':
        if in_range(y - 1, x + 1) and data[y - 1][x + 1] == 'S' and in_range(y + 1, x + 1) and data[y + 1][x + 1] == 'S':
            count += 1

    #right two
    if in_range(y - 1, x + 1) and data[y - 1][x + 1] == 'M' and in_range(y + 1, x + 1) and data[y + 1][x + 1] == 'M':
        if in_range(y - 1, x - 1) and data[y - 1][x - 1] == 'S' and in_range(y + 1, x - 1) and data[y + 1][x - 1] == 'S':
            count += 1

    # bottom two
    if in_range(y + 1, x - 1) and data[y + 1][x - 1] == 'M' and in_range(y + 1, x + 1) and data[y + 1][x + 1] == 'M':
        if in_range(y - 1, x - 1) and data[y - 1][x - 1] == 'S' and in_range(y - 1, x + 1) and data[y - 1][x + 1] == 'S':
            count += 1


for y in range(len(data)):
    for x in range(len(data[y])):
        if data[y][x] == 'A':
            check_x_mas(y, x)


print("#2 ->", count)

File: day_04/day_04.cpp
#include <vector>
#include <tuple>
#include <array>
#include <string>
#include <fstream>
#include <iostream>

std::vector<std::string> readFile(const std::string& filename) {
    std::vector<std::string> data;
    std::ifstream file(filename);
    std::string line;

    if (!file.is_open()) {
        std::cerr << "Error: Could not open file" << std::endl;
        exit(1);
    }
    
    while (std::getline(file, line))
        data.push_back(line);

    file.close();
    return data;
}

bool isInRange(int x, int y, std::vector<std::string> data) {
    return y >= 0 && y < data.size() && \
           x >= 0 && x < data[y].length();
}

// PART 1

std::vector<std::tuple<int, int>> directions_part_1 = {
    {-1, -1}, {+0, -1}, {+1, -1},
    {-1, +0},           {+1, +0},
    {-1, +1}, {+0, +1}, {+1, +1}
};

int checkXMas(int y, int x, int count, std::vector<std::string> data) {
    std::string str = "XMAS";

    for (const auto& [dx, dy] : directions_part_1) {
        int tmpX = x;
        int tmpY = y;
        for (size_t i = 0; i < str.length(); i++) {
            if (!isInRange(tmpX, tmpY, data) || !(str[i] == data[tmpY][tmpX]))
                break;
            tmpX += dx;
            tmpY += dy;
            if (i == str.length() - 1)
                count++;
        }
    }
    return count;
}

// PART 2

std::vector<std::tuple<int, int>> directions_part_2 = {
    {-1, -1}, {+1, -1},
    {-1, +1}, {+1, +1}
};

int checkX_Mas(int y, int x, int count, std::vector<std::string> data) {
    if (data[y][x] != 'A')
        return count;

    int matches = 0;
    for (const auto& [dx, dy] : directions_part_2) {
        if (isInRange(x + dx, y + dy, data) && data[y + dy][x + dx] == 'S' && \
            isInRange(x - dx, y - dy, data) && data[y - dy][x - dx] == 'M')
                    matches++;
    }

    if (matches == 2)
        count++;

    return count;
}

int main() {
    std::vector<std::string> data = readFile("input.txt");
    int count_part_1 = 0;
    int count_part_2 = 0;

    for (size_t y = 0; y < data.size(); y++) {
        for (size_t x = 0; x < data[y].length(); x++) {
            count_part_1 = checkXMas(y, x, count_part_1, data);
            count_part_2 = checkX_Mas(y, x, count_part_2, data);
        }
    }

    std::cout << "#1 -> " << count_part_1 << std::endl;
    std::cout << "#2 -> " << count_part_2 << std::endl;
}